File: leet_GRAPH_1743_topo___BFS.ts
function restoreArray(adjacentPairs: number[][]): number[] {
    let D: { [key: number]: number[] } = {}
    let idg: { [key: number]: number } = {}
    for (const pair of adjacentPairs) {
      const [a, b] = pair
      D[a] = D[a] || []
      D[b] = D[b] || []
      D[a].push(b)
      D[b].push(a)
      idg[a] = (idg[a] || 0) + 1
      idg[b] = (idg[b] || 0) + 1
    }
    let size = adjacentPairs.length + 1
    let res: number[] = []
    let isFound = false
    let head: number, tail: number
    for (let [k, v] of Object.entries(idg)) {
      let key = parseInt(k, 10)
      if (v === 1) {
        if (isFound) {
          tail = key
          break
        } else {
          isFound = true
          head = key
        }
      }
    }
    let dq: number[] = []
    dq.push(head)
    while (dq.length > 0) {
      let node = dq.shift() as number
      res.push(node)
      for (let n of D[node]) {
        idg[n]--
        if (idg[n] === 1) {
          dq.push(n)
        }
      }
    }
    res.push(tail)
    return res
};

File: leet_GRAPH_1743_topo___BFS.go
func restoreArray(adjacentPairs [][]int) []int {
    D := make(map[int][]int)
	idg := make(map[int]int)

	for _, pair := range adjacentPairs {
		a, b := pair[0], pair[1]
		D[a] = append(D[a], b)
		D[b] = append(D[b], a)
		idg[a]++
		idg[b]++
	}

	res := make([]int, 0, size)
	isFound := false

	// look for the beginning
	var head, tail int
	for k, v := range idg {
		if v == 1 {
			if isFound {
				tail = k
				break
			} else {
				isFound = true
				head = k
			}
		}
	}

	dq := []int{head} // slice is enough for bfs like in python
	index := 0

	for index < len(dq) {
		node := dq[index]
		index++
		res = append(res, node)

		for _, n := range D[node] {
			idg[n]--
			if idg[n] == 1 {
				dq = append(dq, n)
			}
		}
	}

	res = append(res, tail)
	return res
}
